src/walker.rs

```diff
@@ -0,0 +1,183 @@
+        ctx: &Context,
+        cur_block: Ptr<BasicBlock>,
+    ) -> Vec<VisitAction<Ptr<BasicBlock>>>;
+
+    /// Given a block, get the list of [Operation]s in the block that must be visited.
+    fn operations_visit_order(
+        &mut self,
+        ctx: &Context,
+        cur_block: Ptr<BasicBlock>,
+    ) -> Vec<Ptr<Operation>>;
+}
+
+/// Callbacks when a [BasicBlock] is processed.
+pub trait ProcessBlock {
+    /// Call back before visiting the instructions in a [BasicBlock].
+    fn process_pre_insts(&mut self, _ctx: &mut Context, _block: Ptr<BasicBlock>) -> Result<()> {
+        Ok(())
+    }
+    /// Call back after visiting the instructions in a [BasicBlock].
+    fn process_post_insts(&mut self, _ctx: &mut Context, _block: Ptr<BasicBlock>) -> Result<()> {
+        Ok(())
+    }
+}
+
+/// When a walk arrives at a [Region], specify which of its blocks, and in what
+/// order, must be visited.
+pub trait VisitRegion {
+    /// Given a region, get a list of child blocks that must be visited.
+    /// Typically, this is just the entry block or the exit blocks since
+    /// the other blocks subsequently get visited via [VisitBlock::blocks_visit_order].
+    fn blocks_visit_order(
+        &mut self,
+        ctx: &Context,
+        cur_region: Ptr<Region>,
+    ) -> Vec<Ptr<BasicBlock>>;
+}
+
+/// Callbacks when a [Region] is processed.
+pub trait ProcessRegion {
+    /// Call back before visiting the [BasicBlock]s in a [Region].
+    fn process_pre_blocks(&mut self, _ctx: &mut Context, _region: Ptr<Region>) -> Result<()> {
+        Ok(())
+    }
+    /// Call back after visiting the [BasicBlock]s in a [Region].
+    fn process_post_blocks(&mut self, _ctx: &mut Context, _region: Ptr<Region>) -> Result<()> {
+        Ok(())
+    }
+}
+
+/// When a walk arrives at an [Operation],
+/// specify the order in which its [Region]s are visited.
+pub trait VisitOperation {
+    /// Given an [Operation], get a list of child [Region]s to be visited, in that order.
+    fn region_visit_order(
+        &mut self,
+        ctx: &Context,
+        cur_operation: Ptr<Operation>,
+    ) -> Vec<Ptr<Region>>;
+}
+
+/// Callbacks when an [Operation] is processed.
+pub trait ProcessOperation {
+    /// Call back before visiting the [Region]s in an [Operation].
+    fn process_pre_regions(
+        &mut self,
+        _ctx: &mut Context,
+        _operation: Ptr<Operation>,
+    ) -> Result<()> {
+        Ok(())
+    }
+    /// Call back after visiting the [Region]s in an [Operation].
+    fn process_post_regions(
+        &mut self,
+        _ctx: &mut Context,
+        _operation: Ptr<Operation>,
+    ) -> Result<()> {
+        Ok(())
+    }
+}
+
+/// Given Visitors and Processors (see [module documentation](crate::walker)),
+/// walk the entire IR graph, starting at a provided root node.
+pub struct Walker<Visitor, Processor>
+where
+    Visitor: VisitBlock + VisitOperation + VisitRegion,
+    Processor: ProcessBlock + ProcessOperation + ProcessRegion,
```

Comment:
>what is the advantage of splitting Processor and Visitor into multiple traits if we have to implement all traits anyways?

I kind of thought that it would help reuse code, but on hindsight, it's probably more cumbersome to split it.

>or do we expect users to rely on the types in the walkers module?

Yes, I was relying on this thought.